src/menu/menu.c

```diff
@@ -856,22 +970,200 @@ menu_open(struct menu *menu, int x, int y)
 	menu->server->input_mode = LAB_INPUT_STATE_MENU;
 }
 
+struct pipe_context {
+	struct server *server;
+	struct menuitem *item;
+	struct buf buf;
+	struct wl_event_source *event_read;
+	struct wl_event_source *event_timeout;
+	pid_t pid;
+	int pipe_fd;
+};
+
+static int
+handle_pipemenu_timeout(void *_ctx)
+{
+	struct pipe_context *ctx = _ctx;
+	kill(ctx->pid, SIGTERM);
+	return 0;
+}
+
+static int
+handle_pipemenu_readable(int fd, uint32_t mask, void *_ctx)
+{
+	struct pipe_context *ctx = _ctx;
+	char data[4096];
```

Comment:
Use `4097` to read a complete page + terminating NULL byte

Reply:
I've now changes this to 2 4k pages + NULL byte.
Basically this controls how much data we read at once from the pipe, it is a trade-off with (temporary) stack memory usage vs minimal increase in latency when we have to wait for the next fd readable event. Especially with app menu pipe menus in mind, we could technically increase this up to 64k which is what a pipe can hold internally by default as far as I remember.

I would be interesting to know how big usual application menu pipe menus are and if they write everything at once or in shorter bursts. Can be seen from the log entries like `00:38:19.699 [../src/menu/menu.c:1103] read 1073 bytes of data`.

Reply:
I'm fine with two pages. There is the BUFSIZ macro - which I think it meant to give an optimum size (4096 generally I believe), but it probably wasn't tuned for our use-case!!

I suspect most pipemenus produce way less than 4096 anyway.

The biggest ones I've seen are similar to `browse.sh` and one that displays all keybinds (by parsing `rc.xml`). But again, even the keybind one shouldn't be that big.